File: swift-tests/NitroLoggerFileWriterTests/AuthContractSupport.swift
import Foundation

enum AuthContractSupport {
  private static func text(_ object: [String: Any], _ key: String) -> String? {
    object[key] as? String
  }

  private static func exactScope(
    _ fields: [String], _ presented: [String: Any]?, _ derived: [String: Any]?
  ) -> Bool {
    guard Set(presented?.keys ?? Dictionary<String, Any>().keys) == Set(fields),
          Set(derived?.keys ?? Dictionary<String, Any>().keys) == Set(fields) else { return false }
    return fields.allSatisfy { field in
      guard let left = presented?[field] as? String,
            let right = derived?[field] as? String else { return false }
      return left == right
    }
  }

  private static let mutations = [
    "capability-mint", "capability-exchange", "manifest-registration", "upload-acceptance",
  ]
  private static let transitions = [
    "subject-deletion", "consent-revocation", "tenant-or-binding-disable",
    "generation-change", "delivery-deadline",
  ]

  private static func validRefusal(_ input: [String: Any]) -> Bool {
    guard text(input, "contentType") == "application/json",
          let body = input["body"] as? [String: Any],
          Set(body.keys) == Set(["contractVersion", "operation", "code"]),
          body["contractVersion"] as? Int == 1,
          body["operation"] as? String == text(input, "requestOperation"),
          body["code"] as? String == "refused" else { return false }
    return true
  }

  static func evaluate(
    contract: [String: Any], group: String, row: [String: Any]
  ) -> String? {
    guard let input = row["input"] as? [String: Any],
          let scope = contract["scope"] as? [String: Any],
          let fields = scope["fields"] as? [String] else { return nil }
    if group == "scope" {
      return exactScope(
        fields, input["presented"] as? [String: Any], input["derived"] as? [String: Any]
      ) ? "allow" : "refuse"
    }
    if group == "mint" {
      if text(input, "operation") == "root-rotation" { return "refuse" }
      if input["newGeneration"] != nil { return "invalidate-old-then-mint" }
      guard ["live-session", "tenant-backend"].contains(text(input, "authority")) else {
        return "refuse"
      }
      if let claim = input["scopeClaim"] as? [String: Any],
         let derived = input["derivedScope"] as? [String: Any],
         !exactScope(fields, claim, derived) { return "refuse" }
      if text(input, "binding") == "consent-revoked" { return "refuse" }
      if let existing = text(input, "existingMintId") {
        return existing == text(input, "mintId") ? "replay-exact-response" : "refuse"
      }
      if input["existingRoot"] as? Bool == false { return "mint-and-record" }
      return "mint-root"
    }
    if group == "exchange" {
      if input["atDeadline"] as? Bool == true || input["afterDeadline"] as? Bool == true {
        return "refuse"
      }
      if ["forged", "restored-consumed"].contains(text(input, "credential")) { return "refuse" }
      if text(input, "binding") == "disabled" { return "refuse" }
      if let recorded = text(input, "recordedExchangeId") {
        return recorded == text(input, "exchangeId") ? "replay-exact-response" : "refuse"
      }
      if input["deliveryDeadline"] != nil {
        return "successor-expires-no-later-than-delivery-deadline"
      }
      return "consume-mint-record"
    }
    if group == "registration" {
      if input["uploads"] != nil { return "exercise-without-consuming" }
      if text(input, "authority") != "live-session" { return "refuse" }
      if let claim = input["scopeClaim"] as? [String: Any],
         let derived = input["derivedScope"] as? [String: Any],
         !exactScope(fields, claim, derived) { return "refuse" }
      if text(input, "record") == "same-id-different-hash" { return "refuse" }
      if text(input, "record") == "same-id-same-hash" { return "idempotent" }
      return "create"
    }
    if group == "responses" {
      let status = input["status"] as? Int
      if status == 403 { return validRefusal(input) ? "terminal" : "indeterminate-retry-same-id" }
      if status == 429 && input["remainingDeadlineSeconds"] as? Int == 0 { return "do-not-attempt" }
      if status == 429 {
        guard let raw = text(input, "retryAfter"), let delay = Int(raw),
              String(delay) == raw, (1...60).contains(delay) else {
          return "indeterminate-local-backoff"
        }
        let remaining = input["remainingDeadlineSeconds"] as? Int ?? delay
        return "retry-after-\(min(delay, remaining))"
      }
      return "indeterminate-retry-same-id"
    }
    if group == "resourceControls" {
      return input["saturated"] as? Bool == true ? "generic-throttle" : "allow"
    }
    if group == "audit" {
      if text(input, "wal") == "full" { return "non-droppable-aggregate-loss-signal" }
      if text(input, "wal") == "available" { return "durable-refusal-event" }
      if text(input, "publication") == "unavailable" { return "queue-and-retry" }
      if input["deliveryAttempts"] != nil { return "one-logical-event" }
      if input["stateCommit"] as? Bool == true && input["auditIntentCommit"] as? Bool == true {
        return "acknowledge"
      }
      return "fail-without-success"
    }
    if group == "lifecycleRaces" {
      guard let mutation = text(input, "mutation"), mutations.contains(mutation),
            let transition = text(input, "transition"), transitions.contains(transition),
            ["mutation", "transition"].contains(text(input, "winner") ?? "") else {
        return "invalid-vector"
      }
      return text(input, "winner") == "mutation"
        ? "commit-then-invalidate" : "no-new-authority-or-success-state"
    }
    if group == "crashRecovery" {
      if text(input, "crashAt") == "before-state-and-audit-commit" {
        return "no-root-no-success-event"
      }
      if text(input, "crashAt") == "after-state-and-audit-commit",
         input["sameMintIdRetry"] as? Bool == true { return "replay-exact-root" }
      if text(input, "crashAt") == "before-predecessor-consume" {
        return "predecessor-remains-live"
      }
      if text(input, "crashAt") == "after-successor-and-audit-commit",
         input["sameExchangeIdRetry"] as? Bool == true { return "replay-exact-successor" }
      if text(input, "crashAt") == "after-inbox-ledger-audit-commit" {
        return "replay-recorded-ack"
      }
      if text(input, "crashAt") == "after-audit-intent-commit" {
        return "publish-idempotently-after-restart"
      }
      return "invalid-vector"
    }
    return nil
  }

  static func resolve(_ input: [String: Any]) -> String {
    if text(input, "credentialOrScope") == "invalid" { return "refuse" }
    if text(input, "ledger") == "same-id-same-hash" { return "replay-recorded-outcome" }
    if text(input, "ledger") == "same-id-different-hash" { return "refuse" }
    if text(input, "binding") != "live" { return "refuse" }
    if text(input, "manifest") != "exact-match" { return "refuse" }
    return "commit-acceptance"
  }
}

File: android/src/test/java/com/margelo/nitro/nitrologger/AuthContractSupport.kt
package com.margelo.nitro.nitrologger

import com.google.gson.JsonObject

object AuthContractSupport {
  private fun text(input: JsonObject, name: String): String? =
    input.get(name)?.takeUnless { it.isJsonNull }?.asString

  private fun exactScope(fields: List<String>, presented: JsonObject?, derived: JsonObject?): Boolean =
    presented?.keySet() == fields.toSet() && derived?.keySet() == fields.toSet() && fields.all { field ->
      presented?.has(field) == true && derived?.has(field) == true &&
        presented.get(field).isJsonPrimitive && presented.get(field).asJsonPrimitive.isString &&
        text(presented, field) == text(derived, field)
    }

  private val mutations = setOf(
    "capability-mint", "capability-exchange", "manifest-registration", "upload-acceptance"
  )
  private val transitions = setOf(
    "subject-deletion", "consent-revocation", "tenant-or-binding-disable",
    "generation-change", "delivery-deadline"
  )

  private fun validRefusal(input: JsonObject): Boolean {
    if (text(input, "contentType") != "application/json" || !input.get("body")?.isJsonObject.orFalse()) return false
    val body = input.getAsJsonObject("body")
    val version = body.get("contractVersion")
    val operation = body.get("operation")
    val code = body.get("code")
    return body.keySet() == setOf("contractVersion", "operation", "code") &&
      version?.isJsonPrimitive == true && version.asJsonPrimitive.isNumber &&
      runCatching { version.asInt }.getOrNull() == 1 &&
      operation?.isJsonPrimitive == true && operation.asJsonPrimitive.isString &&
      operation.asString == text(input, "requestOperation") &&
      code?.isJsonPrimitive == true && code.asJsonPrimitive.isString && code.asString == "refused"
  }

  private fun Boolean?.orFalse() = this == true

  fun evaluate(contract: JsonObject, group: String, row: JsonObject): String? {
    val input = row.getAsJsonObject("input")
    val fields = contract.getAsJsonObject("scope").getAsJsonArray("fields").map { it.asString }
    return when (group) {
      "scope" -> if (exactScope(fields, input.getAsJsonObject("presented"), input.getAsJsonObject("derived"))) "allow" else "refuse"
      "mint" -> when {
        text(input, "operation") == "root-rotation" -> "refuse"
        input.has("newGeneration") -> "invalidate-old-then-mint"
        text(input, "authority") !in setOf("live-session", "tenant-backend") -> "refuse"
        input.has("derivedScope") && !exactScope(fields, input.getAsJsonObject("scopeClaim"), input.getAsJsonObject("derivedScope")) -> "refuse"
        text(input, "binding") == "consent-revoked" -> "refuse"
        input.has("existingMintId") && text(input, "existingMintId") == text(input, "mintId") -> "replay-exact-response"
        input.has("existingMintId") -> "refuse"
        input.has("existingRoot") && !input.get("existingRoot").asBoolean -> "mint-and-record"
        else -> "mint-root"
      }
      "exchange" -> when {
        input.get("atDeadline")?.asBoolean == true || input.get("afterDeadline")?.asBoolean == true -> "refuse"
        text(input, "credential") in setOf("forged", "restored-consumed") -> "refuse"
        text(input, "binding") == "disabled" -> "refuse"
        input.has("recordedExchangeId") && text(input, "recordedExchangeId") == text(input, "exchangeId") -> "replay-exact-response"
        input.has("recordedExchangeId") -> "refuse"
        input.has("deliveryDeadline") -> "successor-expires-no-later-than-delivery-deadline"
        else -> "consume-mint-record"
      }
      "registration" -> when {
        input.has("uploads") -> "exercise-without-consuming"
        text(input, "authority") != "live-session" -> "refuse"
        input.has("derivedScope") && !exactScope(fields, input.getAsJsonObject("scopeClaim"), input.getAsJsonObject("derivedScope")) -> "refuse"
        text(input, "record") == "same-id-different-hash" -> "refuse"
        text(input, "record") == "same-id-same-hash" -> "idempotent"
        else -> "create"
      }
      "responses" -> when {
        input.get("status").asInt == 403 -> if (validRefusal(input)) "terminal" else "indeterminate-retry-same-id"
        input.get("status").asInt == 429 && input.get("remainingDeadlineSeconds")?.asInt == 0 -> "do-not-attempt"
        input.get("status").asInt == 429 -> {
          val raw = text(input, "retryAfter")
          val delay = raw?.toIntOrNull()
          if (delay == null || delay.toString() != raw || delay !in 1..60) "indeterminate-local-backoff"
          else "retry-after-${minOf(delay, input.get("remainingDeadlineSeconds")?.asInt ?: delay)}"
        }
        else -> "indeterminate-retry-same-id"
      }
      "resourceControls" -> if (input.get("saturated")?.asBoolean == true) "generic-throttle" else "allow"
      "audit" -> when {
        text(input, "wal") == "full" -> "non-droppable-aggregate-loss-signal"
        text(input, "wal") == "available" -> "durable-refusal-event"
        text(input, "publication") == "unavailable" -> "queue-and-retry"
        input.has("deliveryAttempts") -> "one-logical-event"
        input.get("stateCommit")?.asBoolean == true && input.get("auditIntentCommit")?.asBoolean == true -> "acknowledge"
        else -> "fail-without-success"
      }
      "lifecycleRaces" -> when {
        text(input, "mutation") !in mutations || text(input, "transition") !in transitions ||
          text(input, "winner") !in setOf("mutation", "transition") -> "invalid-vector"
        text(input, "winner") == "mutation" -> "commit-then-invalidate"
        else -> "no-new-authority-or-success-state"
      }
      "crashRecovery" -> when {
        text(input, "crashAt") == "before-state-and-audit-commit" -> "no-root-no-success-event"
        text(input, "crashAt") == "after-state-and-audit-commit" && input.get("sameMintIdRetry")?.asBoolean == true -> "replay-exact-root"
        text(input, "crashAt") == "before-predecessor-consume" -> "predecessor-remains-live"
        text(input, "crashAt") == "after-successor-and-audit-commit" && input.get("sameExchangeIdRetry")?.asBoolean == true -> "replay-exact-successor"
        text(input, "crashAt") == "after-inbox-ledger-audit-commit" -> "replay-recorded-ack"
        text(input, "crashAt") == "after-audit-intent-commit" -> "publish-idempotently-after-restart"
        else -> "invalid-vector"
      }
      else -> null
    }
  }

  fun resolve(input: JsonObject): String = when {
    text(input, "credentialOrScope") == "invalid" -> "refuse"
    text(input, "ledger") == "same-id-same-hash" -> "replay-recorded-outcome"
    text(input, "ledger") == "same-id-different-hash" -> "refuse"
    text(input, "binding") != "live" -> "refuse"
    text(input, "manifest") != "exact-match" -> "refuse"
    else -> "commit-acceptance"
  }
}
